packages/mobx-state-tree/src/types/utility-types/custom.ts

```diff
@@ -0,0 +1,145 @@
+import {
+    INode,
+    createNode,
+    Type,
+    IType,
+    TypeFlags,
+    IContext,
+    IValidationResult,
+    typeCheckSuccess,
+    typeCheckFailure,
+    ObjectNode
+} from "../../internal"
+
+export type CustomTypeOptions<S, T> = {
+    // Friendly name
+    name: string
+    // given a serialized value, how to turn it into the target type
+    fromSnapshot(snapshot: S): T
+    // return the serialization of the current value
+    toSnapshot(value: T): S
+    // if true, this is a converted value, if false, it's a snapshot
+    isTargetType(value: T | S): boolean
+    // a non empty string is assumed to be a validation error
+    isValidSnapshot(snapshot: S): string
+    // TODO: isSnapshotEqual
+    // TODO: isValueEqual
+}
+
+/**
+ * Creates a custom type. Custom types can be used for arbitrary immutable values, that have a serializable representation. For example, to create your own Date representation, Decimal type etc.
+ *
+ * The signature of the options is:
+ *
+ * ```
+ * export type CustomTypeOptions<S, T> = {
+ *     // Friendly name
+ *     name: string
+ *     // given a serialized value, how to turn it into the target type
+ *     fromSnapshot(snapshot: S): T
+ *     // return the serialization of the current value
+ *     toSnapshot(value: T): S
+ *     // if true, this is a converted value, if false, it's a snapshot
+ *     isTargetType(value: T | S): boolean
+ *     // a non empty string is assumed to be a validation error
+ *     isValidSnapshot?(snapshot: S): string
+ * }
+ * ```
+ *
+ * @export
+ * @alias types.custom
+ *
+ * @example
+ * const DecimalPrimitive = types.custom<string, Decimal>({
+ *     name: "Decimal",
+ *     fromSnapshot(value: string) {
+ *         return new Decimal(value)
+ *     },
+ *     toSnapshot(value: Decimal) {
+ *         return value.toString()
+ *     },
+ *     isTargetType(value: string | Decimal): boolean {
+ *         return value instanceof Decimal
+ *     },
+ *     isValidSnapshot(value: string): string {
+ *         if (/^-?\d+\.\d+$/.test(value)) return "" // OK
+ *         return `'${value}' doesn't look like a valid decimal number`
+ *     }
+ * })
+ *
+ * const Wallet = types.model({
+ *     balance: DecimalPrimitive
+ * })
+ */
+export function custom<S, T>(options: CustomTypeOptions<S, T>): IType<S, T> {
+    return new CustomType(options)
+}
+
+export class CustomType<S, T> extends Type<S, T> {
+    readonly flags = TypeFlags.Reference
+    readonly shouldAttachNode = false
+
+    constructor(protected readonly options: CustomTypeOptions<S, T>) {
+        super(options.name)
+    }
+
+    describe() {
+        return this.name
+    }
+
+    isAssignableFrom(type: IType<any, any>): boolean {
+        return type === this
+    }
+
+    isValidSnapshot(value: any, context: IContext): IValidationResult {
+        if (this.options.isTargetType(value)) return typeCheckSuccess()
+        const typeError: string = this.options.isValidSnapshot(value)
```

Comment:
i think transpiled to js one could return true / false => the typeError within the error message would be `Invalid value for type '${this.name}': true`

also isValidSnapshot does imply that one should return a boolean.

maybe we could make that a bit more expressive to avoid confusion.